# EmKa Academy

## UYARI ! : Lütfen çıktının olduğu kod hücresini çalıştırmayın, aksi takdirde almanız gereken çıktı görüntüsünü kaybedebilirsiniz.

- Evrensel kısaltmaları kullanarak numpy ve pandas kütüphanelerini import ediniz.
- mpg.csv dosyasını okuyup df adıyla değişkene atayın

1) df'in ilk 5 satırına, son 5 satırına, shape'ine, infosuna, nan değerlerine ve describe'ına hızlıca bi göz gezdirin.

In [3]:
# Buraya kod yazmayın !

mpg  cylinders  displacement horsepower  weight  acceleration  model_year   
0  18.0          8         307.0        130    3504          12.0          70  \
1  15.0          8         350.0        165    3693          11.5          70   
2  18.0          8         318.0        150    3436          11.0          70   
3  16.0          8         304.0        150    3433          12.0          70   
4  17.0          8         302.0        140    3449          10.5          70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [4]:
# Buraya kod yazmayın !

mpg  cylinders  displacement horsepower  weight  acceleration   
393  27.0          4         140.0         86    2790          15.6  \
394  44.0          4          97.0         52    2130          24.6   
395  32.0          4         135.0         84    2295          11.6   
396  28.0          4         120.0         79    2625          18.6   
397  31.0          4         119.0         82    2720          19.4   

     model_year  origin             name  
393          82       1  ford mustang gl  
394          82       2        vw pickup  
395          82       1    dodge rampage  
396          82       1      ford ranger  
397          82       1       chevy s-10

In [5]:
# Buraya kod yazmayın !

(398, 9)

In [6]:
# Buraya kod yazmayın !

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   name          398 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.1+ KB


In [7]:
# Buraya kod yazmayın !

mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model_year      0
origin          0
name            0
dtype: int64

In [8]:
# Buraya kod yazmayın !

mpg   cylinders  displacement       weight  acceleration   
count  398.000000  398.000000    398.000000   398.000000    398.000000  \
mean    23.514573    5.454774    193.425879  2970.424623     15.568090   
std      7.815984    1.701004    104.269838   846.841774      2.757689   
min      9.000000    3.000000     68.000000  1613.000000      8.000000   
25%     17.500000    4.000000    104.250000  2223.750000     13.825000   
50%     23.000000    4.000000    148.500000  2803.500000     15.500000   
75%     29.000000    8.000000    262.000000  3608.000000     17.175000   
max     46.600000    8.000000    455.000000  5140.000000     24.800000   

       model_year      origin  
count  398.000000  398.000000  
mean    76.010050    1.572864  
std      3.697627    0.802055  
min     70.000000    1.000000  
25%     73.000000    1.000000  
50%     76.000000    1.000000  
75%     79.000000    2.000000  
max     82.000000    3.000000

2) df'in içindeki sütunların birbirleriyle olan korelasyonlarına(ilişkilerine) bakın. (Sadece numeric sütunların ilişkisini görmek için numeric_only parametresini kullanabilirsiniz)

mpg  cylinders  displacement    weight  acceleration   
mpg           1.000000  -0.775396     -0.804203 -0.831741      0.420289  \
cylinders    -0.775396   1.000000      0.950721  0.896017     -0.505419   
displacement -0.804203   0.950721      1.000000  0.932824     -0.543684   
weight       -0.831741   0.896017      0.932824  1.000000     -0.417457   
acceleration  0.420289  -0.505419     -0.543684 -0.417457      1.000000   
model_year    0.579267  -0.348746     -0.370164 -0.306564      0.288137   
origin        0.563450  -0.562543     -0.609409 -0.581024      0.205873   

              model_year    origin  
mpg             0.579267  0.563450  
cylinders      -0.348746 -0.562543  
displacement   -0.370164 -0.609409  
weight         -0.306564 -0.581024  
acceleration    0.288137  0.205873  
model_year      1.000000  0.180662  
origin          0.180662  1.000000

3) mpg sütunu ile diğer sütunların arasındaki korelasyonları özel olarak inceleyin.

mpg             1.000000
cylinders      -0.775396
displacement   -0.804203
weight         -0.831741
acceleration    0.420289
model_year      0.579267
origin          0.563450
Name: mpg, dtype: float64

4) cylinders sütununda hangi unique değerler olduğuna liste halinde bakın.

array([8, 4, 6, 3, 5], dtype=int64)

5) cylinders sütunundaki unique değerlerin her birinden kaçar adet olduğuna bakın.

cylinders
4    204
8    103
6     84
3      4
5      3
Name: count, dtype: int64

6) Arabanın modeli 75-80 yılları arasındaki arabaların hepsini getirin.(75 ve 80 dahil)

mpg  cylinders  displacement horsepower  weight  acceleration   
152  19.0          6         225.0         95    3264          16.0  \
153  18.0          6         250.0        105    3459          16.0   
154  15.0          6         250.0         72    3432          21.0   
155  15.0          6         250.0         72    3158          19.5   
156  16.0          8         400.0        170    4668          11.5   
..    ...        ...           ...        ...     ...           ...   
333  32.7          6         168.0        132    2910          11.4   
334  23.7          3          70.0        100    2420          12.5   
335  35.0          4         122.0         88    2500          15.1   
336  23.6          4         140.0          ?    2905          14.3   
337  32.4          4         107.0         72    2290          17.0   

     model_year  origin                     name  
152          75       1  plymouth valiant custom  
153          75       1           chevrolet nova  
154          75       1          mercury monarch  
155          75       1            ford maverick  
156          75       1         pontiac catalina  
..          ...     ...                      ...  
333          80       3            datsun 280-zx  
334          80       3            mazda rx-7 gs  
335          80       2        triumph tr7 coupe  
336          80       1       ford mustang cobra  
337          80       3             honda accord  

[186 rows x 9 columns]

7) horsepower'ı 155-180 arasındaki arabaların hepsini getirin.(155 dahil, 180 dahil değil)

not1 : Bu sorguyu yapabilmeniz için horsepower sütununu object durumdan float duruma getirmeniz gerekiyor.

not2 : horsepower sütunu içinde '?' şeklinde değerler var. Bu değerleri önce nan haline getirip devamında horsepower sütununu float hale getirip istenen sorguyu yapabilirsiniz.

not3 : nan değerler varken bile sütunun dtype'ını 'float' yapabilme sebebimiz nan değerlerin arka planda içerisinde rastgele anlamsız float bir sayı tutmasıdır.

mpg  cylinders  displacement  horsepower  weight  acceleration   
1    15.0          8         350.0       165.0    3693          11.5  \
10   15.0          8         383.0       170.0    3563          10.0   
11   14.0          8         340.0       160.0    3609           8.0   
38   14.0          8         350.0       165.0    4209          12.0   
39   14.0          8         400.0       175.0    4464          11.5   
43   13.0          8         400.0       170.0    4746          12.0   
44   13.0          8         400.0       175.0    5140          12.0   
62   13.0          8         350.0       165.0    4274          12.0   
63   14.0          8         400.0       175.0    4385          12.0   
68   13.0          8         350.0       155.0    4502          13.5   
69   12.0          8         350.0       160.0    4456          13.5   
85   13.0          8         350.0       175.0    4100          13.0   
92   13.0          8         351.0       158.0    4363          13.0   
96   13.0          8         360.0       175.0    3821          11.0   
104  12.0          8         400.0       167.0    4906          12.5   
105  13.0          8         360.0       170.0    4654          13.0   
156  16.0          8         400.0       170.0    4668          11.5   
230  15.5          8         350.0       170.0    4165          11.4   
263  17.7          6         231.0       165.0    3445          13.4   
289  16.9          8         350.0       155.0    4360          14.9   

     model_year  origin                             name  
1            70       1                buick skylark 320  
10           70       1              dodge challenger se  
11           70       1               plymouth 'cuda 340  
38           71       1                 chevrolet impala  
39           71       1        pontiac catalina brougham  
43           71       1         ford country squire (sw)  
44           71       1              pontiac safari (sw)  
62           72       1                 chevrolet impala  
63           72       1                 pontiac catalina  
68           72       1             buick lesabre custom  
69           72       1       oldsmobile delta 88 royale  
85           73       1                buick century 350  
92           73       1                         ford ltd  
96           73       1          amc ambassador brougham  
104          73       1                     ford country  
105          73       1           plymouth custom suburb  
156          75       1                 pontiac catalina  
230          77       1     chevrolet monte carlo landau  
263          78       1  buick regal sport coupe (turbo)  
289          79       1          buick estate wagon (sw)

8) Elinizdeki df'i horsepower'a göre büyükten küçüğe sıralayın ardından df'in indexlerini sıfırlayın.

mpg  cylinders  displacement  horsepower  weight  acceleration   
0    26.0          4          97.0        46.0    1835          20.5  \
1    26.0          4          97.0        46.0    1950          21.0   
2    43.4          4          90.0        48.0    2335          23.7   
3    44.3          4          90.0        48.0    2085          21.7   
4    43.1          4          90.0        48.0    1985          21.5   
..    ...        ...           ...         ...     ...           ...   
393  21.0          6         200.0         NaN    2875          17.0   
394  40.9          4          85.0         NaN    1835          17.3   
395  23.6          4         140.0         NaN    2905          14.3   
396  34.5          4         100.0         NaN    2320          15.8   
397  23.0          4         151.0         NaN    3035          20.5   

     model_year  origin                             name  
0            70       2     volkswagen 1131 deluxe sedan  
1            73       2          volkswagen super beetle  
2            80       2               vw dasher (diesel)  
3            80       2             vw rabbit c (diesel)  
4            78       2  volkswagen rabbit custom diesel  
..          ...     ...                              ...  
393          74       1                    ford maverick  
394          80       2             renault lecar deluxe  
395          80       1               ford mustang cobra  
396          81       2                      renault 18i  
397          82       1                   amc concord dl  

[398 rows x 9 columns]

9) horsepower sütunununu float dtype'ına döndürmek için '?' değerlerini nan'a döndürmüştük ancak öyle bırakmıştık. Şimdi bu nan değerleri doldurma vakti.

Basitçe horsepower sütununun ortalamasıyla horsepower içindeki nan değerleri doldurmaya çalışın. 

Devamında tüm sütunları nan değer kaldı mı diye kontrol edin ve ondan sonra df'i yazdırın.

mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model_year      0
origin          0
name            0
dtype: int64

mpg  cylinders  displacement  horsepower  weight  acceleration   
0    26.0          4          97.0   46.000000    1835          20.5  \
1    26.0          4          97.0   46.000000    1950          21.0   
2    43.4          4          90.0   48.000000    2335          23.7   
3    44.3          4          90.0   48.000000    2085          21.7   
4    43.1          4          90.0   48.000000    1985          21.5   
..    ...        ...           ...         ...     ...           ...   
393  21.0          6         200.0  104.469388    2875          17.0   
394  40.9          4          85.0  104.469388    1835          17.3   
395  23.6          4         140.0  104.469388    2905          14.3   
396  34.5          4         100.0  104.469388    2320          15.8   
397  23.0          4         151.0  104.469388    3035          20.5   

     model_year  origin                             name  
0            70       2     volkswagen 1131 deluxe sedan  
1            73       2          volkswagen super beetle  
2            80       2               vw dasher (diesel)  
3            80       2             vw rabbit c (diesel)  
4            78       2  volkswagen rabbit custom diesel  
..          ...     ...                              ...  
393          74       1                    ford maverick  
394          80       2             renault lecar deluxe  
395          80       1               ford mustang cobra  
396          81       2                      renault 18i  
397          82       1                   amc concord dl  

[398 rows x 9 columns]

10) En son indexlerdeki nan olan değerlerin 104.469 değeriyle dolduğunu görüyoruz fakat biz horsepower'a göre sıralamak istemiştik bu df'i. 

Nan değerler doldurulduktan sonra sıralama güncellenmedi. Tekrar horsepower'a göre df'imizi sıralayıp indexleri resetleyelim.

mpg  cylinders  displacement  horsepower  weight  acceleration   
0    26.0          4          97.0        46.0    1835          20.5  \
1    26.0          4          97.0        46.0    1950          21.0   
2    43.4          4          90.0        48.0    2335          23.7   
3    44.3          4          90.0        48.0    2085          21.7   
4    43.1          4          90.0        48.0    1985          21.5   
..    ...        ...           ...         ...     ...           ...   
393  14.0          8         454.0       220.0    4354           9.0   
394  14.0          8         455.0       225.0    4425          10.0   
395  12.0          8         455.0       225.0    4951          11.0   
396  14.0          8         455.0       225.0    3086          10.0   
397  16.0          8         400.0       230.0    4278           9.5   

     model_year  origin                             name  
0            70       2     volkswagen 1131 deluxe sedan  
1            73       2          volkswagen super beetle  
2            80       2               vw dasher (diesel)  
3            80       2             vw rabbit c (diesel)  
4            78       2  volkswagen rabbit custom diesel  
..          ...     ...                              ...  
393          70       1                 chevrolet impala  
394          70       1                 pontiac catalina  
395          73       1         buick electra 225 custom  
396          70       1          buick estate wagon (sw)  
397          73       1               pontiac grand prix  

[398 rows x 9 columns]

11) Elinizdeki df'den en küçük 5 acceralation'a sahip olan arabaları getirin.

mpg  cylinders  displacement  horsepower  weight  acceleration   
360  14.0          8         340.0       160.0    3609           8.0  \
392  14.0          8         440.0       215.0    4312           8.5   
382  15.0          8         390.0       190.0    3850           8.5   
393  14.0          8         454.0       220.0    4354           9.0   
397  16.0          8         400.0       230.0    4278           9.5   

     model_year  origin                name  
360          70       1  plymouth 'cuda 340  
392          70       1   plymouth fury iii  
382          70       1  amc ambassador dpl  
393          70       1    chevrolet impala  
397          73       1  pontiac grand prix

In [ ]:
# Done.